### DISCUSSION TITLE
Modify Partials Feature To Be Compatible With HTMX Recommended Patterns

### DISCUSSION THREAD

**Discussion by codepilotsf** [withastro/roadmap#939] [Proposal]
### Body

# Summary

Partials are currently implemented in a way that is not compatible with recommended patterns in the Hypermedia Systems book (the book written by HTMX creators). 

# The Problem

The recommended pattern prescribes that the same route should be able to respond as either a full page request or as a partial. For example `/books/edit/123` when navigated to directly, should render the populated form within the full page including the layout. But that same URL when navigated to via HTMX, should be able to check for the `HX-Request` header and if present, should render _only_ the populated form as a partial. This currently isn't possible because `export const partial = true` is evaluated at build time instead of run time.

# The Solution

Change the method of designating whether the full document is included to be evaluated at run time for SSR and allow it to be determined even within nested conditionals.

# Also...

Please ALLOW `script` and `style` tags to optionally be included in the partial (not stripped out). The HTML specification allows `script` tags to be present almost anywhere within the DOM. There's no reason it should have to be stripped out and/or moved to the document head. `style` tags are technically not allowed outside the document head (even though it seems to work just fine) but by stripping these out of the partial, the developer loses the ability to use partials and parse the response to move those `style` tags into the head if desired.

**Comment by matthewp** [comment]
I'm not sure if we need to know if a partial is a partial statically or not. It might be possible to determine it after the frontmatter has run. We definitely need to know before the first byte is returned, however.

About scripts/styles; the only meaningful difference between pages and partials is that partials omit these things. So if you want to have scripts and styles then don't set the page to be a partial.

**Reply by codepilotsf** [reply]
Currently, if a template is NOT declared as a partial (at build time), the response will include a `DOCTYPE` tag and if using Tailwind, it will also include all of the compiled Tailwind styles for the whole project. Ideally, a partial could be declared at run time and would only include the content included in that partial. For example:

```
---
// Template can be run as a full page or as a partial (determined at run time)
Astro.partial = Astro.request.headers.get('HX-Request') ? true : false;
---
{ !Astro.partial &&
  <p>include other stuff...</p>
}
<h1>Hello!</h1>

<style>
  h1 { color: green }
</style>
```

In this example, if the template was called as an HTMX request (header is present), we only return the `h1` element and the `style` element. Nothing more. No Doctype. No application-wide Tailwind classes. Just those two elements.

If if was called as a regular, non-HTMX request (header is not present), we return the whole page as normal including the doctype and the Tailwind classes AND also the additional `p` element in this template.

It's worth mentioning that this doesn't need to be a breaking change. The old build-time `export const partial = true` syntax could still be supported (and eventually deprecated).

**Reply by Fryuni** [reply]
The scripts and styles are only removed if they are processed by Astro (which turns scripts into modules and applies scopes to the styles). For both of them, when using `is:inline`, they are not removed from a partial page.

Modules don't re-run if you add them again to the page or add them multiple times. So, having a script to do something once it gets attached by HTMX would not work as one might expect. Attaching them multiple times would be just wasting memory at that point.
I don't know if HTMX handles modules so that they are not attached as duplicates, but that is one point of concern.

Returning the script loading the modules would still be useful in some cases where something just needs to be initialized at some point to use a component. If the component is not being used yet, then there is no need to load the module. Web components are a good example of this use case.

As for styles, I have little to no skills related to CSS, so I don't really understand the difference between processed and non-processed style elements. I know it relates to scoping, but I also don't know what scoping does for styles, so... there's that.

**Reply by codepilotsf** [reply]
Thank you for that helpful info. I'd like to refocus attention on the initial, more pressing need for HTMX devs: The ability to declare partials at runtime. The goal being that for example a route for `GET /users/<id>` should return the full page by default, and _only_ the populated form partial when called with HTMX (after checking for presence of `HX-Request` header at runtime). Currently, we can only set `partial` to `true` at build time. If we omit setting `partial` to `true`, and return only the relevant part of the template (using Astro template conditionals), that _almost_ works -- but not quite -- because that will also return a Doctype (which can break the layout) and if using Tailwind, the response will include a style tag with all of the Tailwind styles for the whole project -- which is very inefficient.

So currently, there's no way to implement the HTMX patterns prescribed in the Hypermedia Systems books using Astro. And in a more general sense, the situation is actually a bit worse than that because: **There is currently no way in Astro to determine at runtime that the response will include only a chunk of DOM without also including all compiled Tailwind classes**. I think this should be addressed urgently as it significantly impedes Astro's utility as an SSR library as a base for a hypermedia-based approach with tools like HTMX, Unpoly, etc.

**Comment by Tc-001** [comment]
https://github.com/withastro/roadmap/pull/901 (once implemented) could be an another quite clean way to solve the issue.
In the partial route you could check if the header does not exist, and reroute to a full page that usually exists on a different path.

**Comment by codepilotsf** [comment]
I'd be interested to know if this can be addressed anytime soon. It is currently impossible to send a response without all my Tailwind styles for the whole site included with every request unless that can be determined at build time. This makes a lot of common patterns that would be trivial if using Express, Flask, Sinatra, whatever –– totally impossible in Astro. If only we could determine at run-time whether the response should be a partial or not, this would be fully solved.

**Comment by codepilotsf** [comment]
This seems to have died on the vine. It's so sad that Astro is very nearly the perfect backend for HTMX -- but doesn't quite work due to the inability to respond with a partial at runtime. It's sooo close! But for now, I guess I have to look at something else like maybe Hono. Bummer :-/

**Reply by matthewp** [reply]
If you want to submit a PR for that I think we would accept.

**Comment by sgalcheung** [comment]
Here is a similar problem https://github.com/withastro/astro/issues/9347
The only difference is that I use Tailwind CSS.
I also submit support there: [Why can't I access some partial directories normally?](https://discord.com/channels/830184174198718474/1273699816123400262)

Is it still a BUG, not a Pull Feature?
